File: backend/app/schemas/broker_chat.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.system_config import LlmProvider

BrokerChatVisibility = Literal["minimal", "tool_calls", "full"]
BrokerChatStatus = Literal["queued", "running", "completed", "failed", "cancelled"]


class BrokerChatRetryPolicyOut(BaseModel):
    enabled: bool = True
    max_retries: int = Field(default=3, ge=0)
    base_delay_seconds: float = Field(default=2.0, ge=0)
    max_delay_seconds: float = Field(default=12.0, ge=0)


class BrokerChatPreferenceOut(BaseModel):
    default_provider: LlmProvider | None = None
    default_model: str | None = None
    event_visibility: BrokerChatVisibility = "full"
    include_tool_outputs: bool = True
    include_reasoning: bool = True
    reasoning_effort: str | None = None
    use_mcp: bool = False
    mcp_server_ids: list[str] = Field(default_factory=list)
    retry: BrokerChatRetryPolicyOut = Field(default_factory=BrokerChatRetryPolicyOut)


class BrokerChatPreferenceUpdateIn(BaseModel):
    default_provider: LlmProvider | None = None
    default_model: str | None = Field(default=None, max_length=256)
    event_visibility: BrokerChatVisibility = "full"
    include_tool_outputs: bool = True
    include_reasoning: bool = True
    reasoning_effort: str | None = None
    use_mcp: bool = False
    mcp_server_ids: list[str] = Field(default_factory=list)
    retry: BrokerChatRetryPolicyOut | None = None


BrokerChatSurface = Literal["broker_chat", "adaptive_workspace"]


class BrokerChatSessionCreateIn(BaseModel):
    title: str | None = Field(default=None, max_length=256)
    surface: BrokerChatSurface | None = None


class BrokerChatSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: str
    title: str
    surface: BrokerChatSurface | str = "broker_chat"
    created_at: datetime
    updated_at: datetime


class BrokerChatSubmitIn(BaseModel):
    message: str = Field(..., min_length=1)
    session_id: str | None = None
    session_title: str | None = Field(default=None, max_length=256)
    provider: LlmProvider | None = None
    model: str | None = Field(default=None, max_length=256)
    event_visibility: BrokerChatVisibility | None = None
    include_tool_outputs: bool | None = None
    include_reasoning: bool | None = None
    reasoning_effort: str | None = None
    use_mcp: bool | None = None
    mcp_server_ids: list[str] | None = None
    default_account_id: str | None = None
    search_account_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class BrokerChatRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    session_id: str
    user_id: str
    status: BrokerChatStatus | str
    job_id: str | None = None
    provider: str
    model_id: str
    message: str
    response_text: str = ""
    error: str | None = None
    event_visibility: BrokerChatVisibility | str = "minimal"
    include_tool_outputs: bool = False
    include_reasoning: bool = False
    metadata_json: str = "{}"
    queued_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class BrokerChatSubmitOut(BaseModel):
    run: BrokerChatRunOut
    stream_url: str
    status_url: str
    events_url: str


class BrokerChatEventOut(BaseModel):
    id: str
    run_id: str
    sequence: int
    event_type: str
    payload: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class BrokerChatEventsPageOut(BaseModel):
    run: BrokerChatRunOut
    events: list[BrokerChatEventOut] = Field(default_factory=list)
    next_after_sequence: int | None = None

File: frontend/service/types/broker-chat.ts
import type { LlmProvider } from "@/service/types/broker";

export type BrokerChatVisibility = "minimal" | "tool_calls" | "full";
export type BrokerChatStatus = "queued" | "running" | "completed" | "failed" | "cancelled" | string;

export type BrokerChatPayload = Record<string, unknown>;

export interface BrokerChatPreference {
    default_provider?: LlmProvider | null;
    default_model?: string | null;
    event_visibility: BrokerChatVisibility;
    include_tool_outputs: boolean;
    include_reasoning: boolean;
    use_mcp: boolean;
}

export interface BrokerChatPreferenceUpdate {
    default_provider?: LlmProvider | null;
    default_model?: string | null;
    event_visibility: BrokerChatVisibility;
    include_tool_outputs: boolean;
    include_reasoning: boolean;
    use_mcp: boolean;
}

export interface BrokerChatSession {
    id: string;
    user_id: string;
    title: string;
    created_at: string;
    updated_at: string;
}

export interface BrokerChatRun {
    id: string;
    session_id: string;
    user_id: string;
    status: BrokerChatStatus;
    job_id?: string | null;
    provider: string;
    model_id: string;
    message: string;
    response_text: string;
    error?: string | null;
    event_visibility: BrokerChatVisibility | string;
    include_tool_outputs: boolean;
    include_reasoning: boolean;
    metadata_json: string;
    queued_at: string;
    started_at?: string | null;
    completed_at?: string | null;
    created_at: string;
    updated_at: string;
}

export interface BrokerChatSubmitRequest {
    message: string;
    session_id?: string | null;
    session_title?: string | null;
    provider?: LlmProvider | null;
    model?: string | null;
    event_visibility?: BrokerChatVisibility | null;
    include_tool_outputs?: boolean | null;
    include_reasoning?: boolean | null;
    use_mcp?: boolean | null;
    default_account_id?: string | null;
    search_account_id?: string | null;
    metadata?: BrokerChatPayload;
}

export interface BrokerChatSubmitResponse {
    run: BrokerChatRun;
    stream_url: string;
    status_url: string;
    events_url: string;
}

export interface BrokerChatEvent {
    id: string;
    run_id: string;
    sequence: number;
    event_type: string;
    payload: BrokerChatPayload;
    created_at: string;
}

export interface BrokerChatEventsPage {
    run: BrokerChatRun;
    events: BrokerChatEvent[];
    next_after_sequence?: number | null;
}

export interface BrokerChatQueueHealth {
    queue_name: string;
    queued_count: number;
    active_worker_count: number;
    has_active_worker: boolean;
    in_process_worker_enabled: boolean;
    has_processing_path: boolean;
    workers: Array<{
        name?: string;
        state?: string;
        queues?: string[];
    }>;
}
